jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java

```diff
@@ -0,0 +1,210 @@
+     * @param graph the chordal graph for which a decomposition should be created
+     */
+    public ChordalNiceDecompositionBuilder(Graph<V, E> graph)
+    {
+        super();
+        ChordalityInspector<V, E> inspec = new ChordalityInspector<>(graph);
+        if (!inspec.isChordal())
+            throw new IllegalArgumentException("The given graph is not chordal.");
+        this.graph = graph;
+        this.perfectOrder = inspec.getPerfectEliminationOrder();
+        vertexInOrder = getVertexInOrder();
+        computeNiceDecomposition();
+    }
+
+    /**
+     * Constructor for the nice decomposition builder of chordal graphs. 
+     * This method needs the perfect elimination order. The constructor
+     * already computes the decomposition.<p>
+     * Note: This method does NOT check whether the order is correct.
+     * 
+     * @param graph the chordal graph for which a decomposition should be created
+     * @param perfectEliminationOrder the perfect elimination order of the graph
+     */
+    public ChordalNiceDecompositionBuilder(Graph<V, E> graph, List<V> perfectEliminationOrder)
+    {
+        super();
+        this.graph = graph;
+        this.perfectOrder = perfectEliminationOrder;
+        computeNiceDecomposition();
+    }
+
+    /**
+     * Computes the nice decomposition of the graph. We iterate over the perfect elimination order
+     * of the chordal graph and try to add a node containing the predecessors regarding the
+     * perfect elimination as a bag to the tree
+     */
+    private void computeNiceDecomposition()
+    {
+        // map from vertices to decomposition-nodes where the decomposition-node has all its
+        // predecessors
+        Map<V, Integer> forgetNodeMap = new HashMap<>(graph.vertexSet().size());
+
+        //iterate over the perfect order
+        Iterator<V> iterator = perfectOrder.iterator();
+        
+        //empty graph
+        if(!iterator.hasNext())
+            return;
+        
+        // set current node to the root
+        V element = iterator.next();
+        int decompNode = addRootNode(element);
+        forgetNodeMap.put(element, decompNode);
+
+        // iterate over the perfect elimination order
+        while (iterator.hasNext()) {
+            V vertex = iterator.next();
+            // get the predecessors regarding the perfect elimination order
+            List<V> predecessors = getOrderPredecessors(vertexInOrder, vertex);
+
+            // calculate nearest successors according to perfect elimination order
+            V lastVertex = null;
+            for (V predecessor : predecessors) {
+                if (lastVertex == null)
+                    lastVertex = predecessor;
+                if (vertexInOrder.get(predecessor) > vertexInOrder.get(lastVertex))
+                    lastVertex = predecessor;
+            }
+
+            // get node with clique of last vertex, else we use the last node
```

Comment:
If the current vertex has no predecessors, then we can use any decomposition node to build a path to, can't we?

Reply:
Correct. Thats why I implicial just use the last one here.